tests: refactor e2e

phisco opened the pull request:
<!--
Thank you for helping to improve Crossplane!

Please read through https://git.io/fj2m9 if this is your first time opening a
Crossplane pull request. Find us in https://slack.crossplane.io/messages/dev if
you need any help contributing.
-->

### Description of your changes

<!--
Briefly describe what this pull request does, and how it is covered by tests.
Be proactive - direct your reviewers' attention to anything that needs special
consideration.

You MUST either [x] check or ~strikethrough~ every item in the checklist below.

We love pull requests that fix an open issue. If yours does, use the below line
to indicate which issue it fixes, for example "Fixes #500".
-->

Fixes #4264.
Refactoring almost all current e2e tests according to the following idea:
- if table testing applies, use `features.Table` to define self-contained assessments, instead of a list of steps
- otherwise stick to 1 feature per Test function using the fluent api, naming all steps with a short CamelCase name, trying to separate setup, assessment and teardown steps separate.

I have:

- [x] Read and followed Crossplane's [contribution process].
- [x] Added or updated unit **and** E2E tests for my change.
- [x] Run `make reviewable` to ensure this PR is ready for review.
- [x] ~Added `backport release-x.y` labels to auto-backport this PR if necessary.~

[contribution process]: https://git.io/fj2m9

phisco committed:
tests: refactor e2e to be more idiomatic

Signed-off-by: Philippe Scorsolini <p.scorsolini@gmail.com>

phisco committed:
tests: set helm timeout and wait through options

Signed-off-by: Philippe Scorsolini <p.scorsolini@gmail.com>

phisco committed:
tests: refactor lifecycle test

Signed-off-by: Philippe Scorsolini <p.scorsolini@gmail.com>

phisco committed:
tests: use reflection to get Kind if empty not

Signed-off-by: Philippe Scorsolini <p.scorsolini@gmail.com>

phisco committed:
tests: minor refactor Upgrade test

Signed-off-by: Philippe Scorsolini <p.scorsolini@gmail.com>

phisco committed:
tests: allow running e2es against existing installation

Signed-off-by: Philippe Scorsolini <p.scorsolini@gmail.com>

phisco committed:
docs: update e2e guidelines

Signed-off-by: Philippe Scorsolini <p.scorsolini@gmail.com>

negz commented on test/e2e/README.md:
````diff
@@ -38,14 +38,22 @@ E2E_TEST_FLAGS="-test.run ^TestConfiguration" make e2e
+E2E_TEST_FLAGS="-test.v -v 4 -test.failfast -destroy-kind-cluster=false -kind-cluster-name=kind -install-crossplane=false -feature=CrossplaneUpgrade" make e2e
+
+# Run the all tests not installing or upgrading Crossplane against the currently selected cluster where Crossplane has already been installed.
+E2E_TEST_FLAGS="-test.v -v 4 -test.failfast -kubeconfig=$HOME/.kube/config -skip-labels install-crossplane=true -create-kind-cluster=false -install-crossplane=false" make go.build e2e-run-tests```
````
It would be neat if we could just infer that if `-install-crossplane=false` then `-skip-labels` should also include `install-crossplane=true`.

phisco replied:
I thought about that, but what if you don't want to install crossplane during the initial spin up, but you want to test the upgrade nonetheless? 🤔 I agree though that the configuration is already becoming a little bit too complex and we should rethink it maybe down the line.

negz commented on test/e2e/main_test.go:
```diff
@@ -37,6 +39,14 @@ import (
 
+// LabelInstallCrossplane is used to mark tests that are going to install
+// Crossplane.
+const LabelInstallCrossplane = "install-crossplane"
```
Would something like "invokes Helm" be clearer here? It's a little surprising that "Installs Crossplane" also includes the cases where we Helm upgrade to turn on a feature flag.

phisco replied:
I wasn't too sure about the name either. It's actually mostly about modifying the crossplane installation's status (installing/upgrading/uninstalling), so I wouldn't specifically mention helm maybe, as it's just an implementation detail.
Maybe "modify-crossplane-config" or "modify-crossplane-installation"? It would make it also less redundant with the "install-crossplane" flag above.

negz replied:
I like `modify-crossplane-installation`.

negz reviewed (commented):
Thanks for iterating on this @phisco! A bunch of comments but nothing really blocking from my end.

phisco commented on test/e2e/README.md:
```diff
@@ -66,10 +74,94 @@ uses a matrix strategy to invoke each area as its own job, running in parallel.
+   needed.
+1. Prefer using the Fluent APIs to define features
+   (`features.New(...).WithSetup(...).Assess(...).WithTeardown(...).Feature()`).
+   1. `features.Table` should be used only to define multiple self-contained
+      assessments to be run sequentially, but without assuming any ordering among
+      them, similarly to the usual table driven style we adopt for unit testing.
```
It's not about the number of assessments, it's about the kind of assessments as we discussed. Table testing can be used when the order of the test cases (that here are assessments) is not relevant and therefore each assessment is self-contained and does not rely nor can be influenced by other assessments.
See the validation test or the example in the e2e folder as an example of an idiomatic table driven test.

negz replied:
I see. And to make sure I understand, this isn't because a `features.Table` is actually run in random order in practice but rather because we think folks would _expect_ that behaviour - i.e. expect a table driven test to not be self/contained and not order dependent.

phisco replied:
Yes, exactly, it's not, but that's what people expect by table driven testing.

phisco commented on test/e2e/README.md:
```diff
@@ -66,10 +74,94 @@ uses a matrix strategy to invoke each area as its own job, running in parallel.
+
+1. Define a single feature per Test function, if possible.
+1. Setup a directory of plain YAML manifests per test - i.e. test fixtures - at
+   `e2e/manifests/<area>/<test>`, usually with a `prerequisites` sub-folder
```
sure!

phisco committed:
chore: address review comments

Signed-off-by: Philippe Scorsolini <p.scorsolini@gmail.com>

phisco commented on test/e2e/apiextensions_test.go:
```diff
@@ -34,213 +34,125 @@ import (
-	// Test that a claim using a very minimal Composition (with no patches,
-	// transforms, or functions) will become available when its composed
-	// resources do.
+// TestCompositionMinimal tests Crossplane's Composition functionality,
+// checking that a claim using a very minimal Composition (with no patches,
+// transforms, or functions) will become available when its composed
+// resources do.
```
I couldn't find the linter to do that though 🤔

negz replied:
The linter that does it is "revive", which is the replacement of the old "golint" linter. My guess is that one of the testing filters is maybe stopping it from being effective on test files.

https://github.com/crossplane/crossplane/blob/3afae0248fb04597ba1feaed3e9fa0fcd8e8ed42/.golangci.yml#L121
https://github.com/crossplane/crossplane/blob/3afae0248fb04597ba1feaed3e9fa0fcd8e8ed42/.golangci.yml#L146

Not something we need to address in this PR though.

phisco commented on test/e2e/README.md:
```diff
@@ -66,10 +74,94 @@ uses a matrix strategy to invoke each area as its own job, running in parallel.
+1. Prefer the usage of `WithSetup` and `WithTeardown` to their unnamed
+   counterparts (`Setup` and `Teardown`) to define the setup and teardown phases of
+   a feature, as they allow to provide a description.
+1. Use short but explicative `CamelCase` sentences as descriptions for
+   everything used to define the name of tests/subtests, e.g.
+   `features.New("CrossplaneUpgrade", ...)` `WithSetup("InstallProviderNop",
+   ...)`, `Assess("ProviderNopIsInstalled", ...)`,
+   `WithTeardown("UninstallProviderNop", ...)`.
```
😆 it's almost identical in Italian ("esplicativo"), "descriptive" would be a more explicative word here? 🤣

phisco commented on test/e2e/README.md:
````diff
@@ -38,14 +38,22 @@ E2E_TEST_FLAGS="-test.run ^TestConfiguration" make e2e
+E2E_TEST_FLAGS="-test.v -v 4 -test.failfast -destroy-kind-cluster=false -kind-cluster-name=kind -install-crossplane=false -feature=CrossplaneUpgrade" make e2e
+
+# Run the all tests not installing or upgrading Crossplane against the currently selected cluster where Crossplane has already been installed.
+E2E_TEST_FLAGS="-test.v -v 4 -test.failfast -kubeconfig=$HOME/.kube/config -skip-labels install-crossplane=true -create-kind-cluster=false -install-crossplane=false" make go.build e2e-run-tests```
````
yes, they get cut off also on my monitor, but you can scroll right to see the rest of the content. However I agree we could wrap them around. let me give it a try

phisco replied:
have a look at the latest change [here](https://github.com/crossplane/crossplane/pull/4270/commits/213edd6fa6091364ce803af36e2886f2a9a80727)

phisco commented on test/e2e/pkg_test.go:
```diff
@@ -32,140 +32,89 @@ import (
-	// a package pull secret.
+// TestConfigurationPullFromPrivateRegistry tests that a Configuration can be
+// installed from a private registry using a package pull secret.
+func TestConfigurationPullFromPrivateRegistry(t *testing.T) {
```
@negz, both this and the test below have a `configuration/crossplane.yaml` file defined, which looks like wasn't used even on master, can we delete them or am I missing something? I wonder if we could think of something to enforce that all fixtures are actually used to avoid having them drift away unnoticed

phisco replied:
Obviously immediately writing this comment I understood the reason of those files 👍 They were actually used to build the Configuration used in the test 👍

negz reviewed (approved):
I think most folks are onboard with these changes. @phisco if you feel good about the implementation I suggest we merge ASAP. We can fast-follow with further improvements. I know at least @pedjak and myself are working on changes that will add E2E tests so it would be ideal to be able to target this new style (and avoid too many merge conflicts).

phisco committed:
chore: rename test label

Signed-off-by: Philippe Scorsolini <p.scorsolini@gmail.com>

phisco committed:
chore: final refactor

Signed-off-by: Philippe Scorsolini <p.scorsolini@gmail.com>

phisco commented:
@negz, totally agree, I've rebased, checked I didn't miss anything porting the tests and renamed a few steps to be aligned with the guidelines defined. Once CI is green it's good to go for me 🙏
